Does mvn's dokka:javadoc download a bunch of stuff?

![Screenshot from 2020-08-30 19-08-40](https://user-images.githubusercontent.com/2416031/91671486-219c4d00-eaf5-11ea-81c3-b520e2d74024.png)

This is my resources during a build.  You can see that around 30 seconds ago, it switched from CPU intensive, to a solid download.  This was the exact point where the `dokka-maven-plugin:1.4.0-rc:javadoc` task was being executed by maven.  It didn't do this in gradle.

```text
$ mvn --version
Apache Maven 3.6.0
Maven home: /usr/share/maven
Java version: 11.0.8, vendor: Ubuntu, runtime: /usr/lib/jvm/java-11-openjdk-amd64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "4.15.0-112-generic", arch: "amd64", family: "unix"
```

```xml
<plugin>
	<groupId>org.jetbrains.dokka</groupId>
	<artifactId>dokka-maven-plugin</artifactId>
	<version>1.4.0-rc</version>
	<executions>
		<execution>
			<phase>pre-site</phase>
			<goals>
				<goal>dokka</goal>
			</goals>
		</execution>
	</executions>
</plugin>
```

Huh.  I eventually waited maybe 4 or 5 minutes and it completed.  Rebuilding again, it only downloaded for 10 or 20 seconds.  What's it doing?  Why?  You might want to add some maven output like, "Downloading the flux capacator...  23% complete" or something.  This is a mixed Java/Kotlin/Maven project.

Another aspect of this is it actually prevents you from compiling without an internet connection.  I was working on a project today, compiling repeatedly (with Gradle) and publishing to mavenLocal.  At one point I had to turn off my network connection and was shocked to discover that I couldn't compile and publish to mavenLocal because that required dokka which failed the build with an `java.net.UnknownHostException: kotlinlang.org` like this:

```
$ gradle clean publishToMavenLocal
Starting a Gradle Daemon, 1 stopped Daemon could not be reused, use --status for details

> Task :dokkaJavadoc
Dokka 1.4.* is an alpha project
Initializing plugins
Creating documentation models
Transforming documentation model before merging
Merging documentation models
Transforming documentation model after merging
Creating pages
Transforming pages
Rendering


> Task :dokkaJavadoc FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':dokkaJavadoc'.
> java.net.UnknownHostException: kotlinlang.org

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 6s
2 actionable tasks: 2 executed
```

It's normal for gradle/maven to download stuff the first time you compile something, but I've never had another JVM project require internet access for subsequent builds.  I'd like to propose that we change this from a "question" to a "bug."  I'm sure I'm not the only person who will be surprised by this behavior and stranded somewhere with no internet and be unable to run a build.  Well, without editing their build file.

I really appreciate having Dokka to generate javadoc-like documentation for Kotlin.  I'm especially glad that the new version works with later versions of the JVM - that's great!  I can see you've fixed and closed a number of bugs already which is excellent.  Thank you for dokka!

--- a/buildSrc/src/main/kotlin/org/jetbrains/SetupMaven.kt
+++ b/buildSrc/src/main/kotlin/org/jetbrains/SetupMaven.kt
@@ -17,9 +17,6 @@ open class SetupMaven : Sync() {
     @get:Input
     var mavenPluginToolsVersion = "3.5.2"
 
-    @get:Input
-    var aetherVersion = "1.1.0"
-
     @get:Internal
     val mavenBuildDir = "${project.buildDir}/maven"
 

--- a/runners/maven-plugin/build.gradle.kts
+++ b/runners/maven-plugin/build.gradle.kts
@@ -11,13 +11,6 @@ dependencies {
     implementation("org.apache.maven.plugin-tools:maven-plugin-annotations:${setupMaven.mavenPluginToolsVersion}")
     implementation("org.apache.maven:maven-archiver:2.5")
     implementation(kotlin("stdlib-jdk8"))
-    implementation("org.eclipse.aether:aether-api:${setupMaven.aetherVersion}")
-    implementation("org.eclipse.aether:aether-spi:${setupMaven.aetherVersion}")
-    implementation("org.eclipse.aether:aether-impl:${setupMaven.aetherVersion}")
-    implementation("org.eclipse.aether:aether-connector-basic:${setupMaven.aetherVersion}")
-    implementation("org.eclipse.aether:aether-transport-file:${setupMaven.aetherVersion}")
-    implementation("org.eclipse.aether:aether-transport-http:${setupMaven.aetherVersion}")
-    implementation("org.apache.maven:maven-aether-provider:3.3.3")
 }
 
 tasks.named<Delete>("clean") {

--- a/runners/maven-plugin/src/main/kotlin/DokkaMojo.kt
+++ b/runners/maven-plugin/src/main/kotlin/DokkaMojo.kt
@@ -2,6 +2,10 @@ package org.jetbrains.dokka.maven
 
 import org.apache.maven.archiver.MavenArchiveConfiguration
 import org.apache.maven.archiver.MavenArchiver
+import org.apache.maven.artifact.DefaultArtifact
+import org.apache.maven.artifact.handler.DefaultArtifactHandler
+import org.apache.maven.artifact.resolver.ArtifactResolutionRequest
+import org.apache.maven.artifact.resolver.ArtifactResolutionResult
 import org.apache.maven.execution.MavenSession
 import org.apache.maven.model.Dependency
 import org.apache.maven.plugin.AbstractMojo
@@ -9,30 +13,15 @@ import org.apache.maven.plugin.MojoExecutionException
 import org.apache.maven.plugins.annotations.*
 import org.apache.maven.project.MavenProject
 import org.apache.maven.project.MavenProjectHelper
-import org.apache.maven.repository.internal.MavenRepositorySystemUtils
+import org.apache.maven.repository.RepositorySystem
 import org.codehaus.plexus.archiver.Archiver
 import org.codehaus.plexus.archiver.jar.JarArchiver
 import org.codehaus.plexus.util.xml.Xpp3Dom
-import org.eclipse.aether.DefaultRepositorySystemSession
-import org.eclipse.aether.RepositorySystem
-import org.eclipse.aether.RepositorySystemSession
-import org.eclipse.aether.artifact.DefaultArtifact
-import org.eclipse.aether.collection.CollectRequest
-import org.eclipse.aether.connector.basic.BasicRepositoryConnectorFactory
-import org.eclipse.aether.graph.DependencyNode
-import org.eclipse.aether.impl.DefaultServiceLocator
-import org.eclipse.aether.repository.LocalRepository
-import org.eclipse.aether.repository.RemoteRepository
-import org.eclipse.aether.resolution.DependencyRequest
-import org.eclipse.aether.spi.connector.RepositoryConnectorFactory
-import org.eclipse.aether.spi.connector.transport.TransporterFactory
-import org.eclipse.aether.transport.file.FileTransporterFactory
-import org.eclipse.aether.transport.http.HttpTransporterFactory
-import org.eclipse.aether.util.graph.visitor.PreorderNodeListGenerator
 import org.jetbrains.dokka.*
 import org.jetbrains.dokka.DokkaConfiguration.ExternalDocumentationLink
 import java.io.File
 import java.net.URL
+import java.util.stream.Collectors
 
 class SourceLinkMapItem {
     @Parameter(name = "path", required = true)
@@ -58,11 +47,18 @@ class ExternalDocumentationLinkBuilder {
 
 abstract class AbstractDokkaMojo(private val defaultDokkaPlugins: List<Dependency>) : AbstractMojo() {
 
-    @Parameter(defaultValue = "\${project}", readonly = true)
-    private var mavenProject: MavenProject? = null
+    @Parameter(defaultValue = "\${project}", readonly = true, required = true)
+    protected var mavenProject: MavenProject? = null
 
-    @Parameter()
-    private var session: RepositorySystemSession? = null
+    /**
+     * The current build session instance. This is used for
+     * for dependency resolver API calls via repositorySystem.
+     */
+    @Parameter(defaultValue = "\${session}", required = true, readonly = true)
+    protected var session: MavenSession? = null
+
+    @Component
+    private var repositorySystem: RepositorySystem? = null;
 
     class PackageOptions : DokkaConfiguration.PackageOptions {
         @Parameter
@@ -254,52 +250,27 @@ abstract class AbstractDokkaMojo(private val defaultDokkaPlugins: List<Dependenc
         gen.generate()
     }
 
-    private fun newRepositorySystem(): RepositorySystem {
-        val locator: DefaultServiceLocator = MavenRepositorySystemUtils.newServiceLocator()
-        locator.addService(RepositoryConnectorFactory::class.java, BasicRepositoryConnectorFactory::class.java)
-        locator.addService(TransporterFactory::class.java, FileTransporterFactory::class.java)
-        locator.addService(TransporterFactory::class.java, HttpTransporterFactory::class.java)
-        return locator.getService(RepositorySystem::class.java)
-    }
-
-    private fun newSession(system: RepositorySystem): RepositorySystemSession {
-        val session: DefaultRepositorySystemSession =
-            MavenRepositorySystemUtils.newSession()
-        val localRepo = LocalRepository(System.getProperty("user.home") + "/.m2/repository")
-        session.localRepositoryManager = system.newLocalRepositoryManager(session, localRepo)
-        return session
-    }
-
     private fun getArtifactByAether(
         groupId: String,
         artifactId: String,
         version: String
     ): List<File> {
-        val repoSystem: RepositorySystem = newRepositorySystem()
-        val session: RepositorySystemSession = newSession(repoSystem)
-        val dependency =
-            org.eclipse.aether.graph.Dependency(DefaultArtifact("$groupId:$artifactId:$version"), "compile")
-        val collectRequest = CollectRequest()
-        collectRequest.root = dependency
-        val repositories: List<RemoteRepository> =
-            (mavenProject?.remoteProjectRepositories?.plus(mavenProject?.remotePluginRepositories ?: emptyList())
-                ?: mavenProject?.remotePluginRepositories ?: emptyList())
-        repositories.forEach {
-            collectRequest.addRepository(
-                RemoteRepository.Builder(
-                    "repo",
-                    "default",
-                    it.url
-                ).setAuthentication( it.authentication ).build()
-            )
-        }
-        val node: DependencyNode = repoSystem.collectDependencies(session, collectRequest).root
-        val dependencyRequest = DependencyRequest()
-        dependencyRequest.root = node
-        repoSystem.resolveDependencies(session, dependencyRequest)
-        val nlg = PreorderNodeListGenerator()
-        node.accept(nlg)
-        return nlg.files
+
+        val request: ArtifactResolutionRequest = ArtifactResolutionRequest()
+        request.isResolveRoot = true
+        request.isResolveTransitively = true
+        request.localRepository = session!!.localRepository
+        request.remoteRepositories = mavenProject!!.pluginArtifactRepositories
+        request.isOffline = session!!.isOffline
+        request.isForceUpdate = session!!.request.isUpdateSnapshots
+        request.servers = session!!.request.servers
+        request.mirrors = session!!.request.mirrors
+        request.proxies = session!!.request.proxies
+        request.artifact = DefaultArtifact(groupId, artifactId, version, "compile", "jar", null,
+                DefaultArtifactHandler("jar"))
+
+        val result: ArtifactResolutionResult = repositorySystem!!.resolve(request)
+        return result.artifacts.stream().map { it.file }.collect(Collectors.toList())
     }
 
     private val dokkaVersion: String by lazy {
@@ -376,12 +347,6 @@ class DokkaJavadocJarMojo : AbstractDokkaMojo(listOf(javadocDependency)) {
     @Parameter(property = "maven.javadoc.classifier", defaultValue = "javadoc", required = true)
     private var classifier: String? = null
 
-    @Parameter(defaultValue = "\${session}", readonly = true, required = true)
-    protected var session: MavenSession? = null
-
-    @Parameter(defaultValue = "\${project}", readonly = true, required = true)
-    protected var project: MavenProject? = null
-
     @Component
     private var projectHelper: MavenProjectHelper? = null
 
@@ -398,7 +363,7 @@ class DokkaJavadocJarMojo : AbstractDokkaMojo(listOf(javadocDependency)) {
         }
         val outputFile = generateArchive("$finalName-$classifier.jar")
         if (attach) {
-            projectHelper?.attachArtifact(project, "javadoc", classifier, outputFile)
+            projectHelper?.attachArtifact(mavenProject, "javadoc", classifier, outputFile)
         }
     }
 
@@ -411,7 +376,7 @@ class DokkaJavadocJarMojo : AbstractDokkaMojo(listOf(javadocDependency)) {
         archiver.archiver.addDirectory(File(outputDir), arrayOf("**/**"), arrayOf())
 
         archive.isAddMavenDescriptor = false
-        archiver.createArchive(session, project, archive)
+        archiver.createArchive(session, mavenProject, archive)
 
         return javadocJar
     }

--- a/runners/maven-plugin/src/main/kotlin/DokkaMojo.kt
+++ b/runners/maven-plugin/src/main/kotlin/DokkaMojo.kt
@@ -53,16 +53,16 @@ abstract class AbstractDokkaMojo(private val defaultDokkaPlugins: List<Dependenc
 
     /**
      * The current build session instance. This is used for
-     * for dependency resolver API calls via repositorySystem.
+     * dependency resolver API calls via repositorySystem.
      */
     @Parameter(defaultValue = "\${session}", required = true, readonly = true)
     protected var session: MavenSession? = null
 
     @Component
-    private var repositorySystem: RepositorySystem? = null;
+    private var repositorySystem: RepositorySystem? = null
 
     @Component
-    private var resolutionErrorHandler: ResolutionErrorHandler? = null;
+    private var resolutionErrorHandler: ResolutionErrorHandler? = null
 
     class PackageOptions : DokkaConfiguration.PackageOptions {
         @Parameter
@@ -260,24 +260,25 @@ abstract class AbstractDokkaMojo(private val defaultDokkaPlugins: List<Dependenc
         version: String
     ): List<File> {
 
-        val request = ArtifactResolutionRequest()
-        request.isResolveRoot = true
-        request.isResolveTransitively = true
-        request.localRepository = session!!.localRepository
-        request.remoteRepositories = mavenProject!!.pluginArtifactRepositories
-        request.isOffline = session!!.isOffline
-        request.isForceUpdate = session!!.request.isUpdateSnapshots
-        request.servers = session!!.request.servers
-        request.mirrors = session!!.request.mirrors
-        request.proxies = session!!.request.proxies
-        request.artifact = DefaultArtifact(groupId, artifactId, version, "compile", "jar", null,
-                DefaultArtifactHandler("jar"))
+        val request = ArtifactResolutionRequest().apply {
+            isResolveRoot = true
+            isResolveTransitively = true
+            localRepository = session!!.localRepository
+            remoteRepositories = mavenProject!!.pluginArtifactRepositories
+            isOffline = session!!.isOffline
+            isForceUpdate = session!!.request.isUpdateSnapshots
+            servers = session!!.request.servers
+            mirrors = session!!.request.mirrors
+            proxies = session!!.request.proxies
+            artifact = DefaultArtifact(groupId, artifactId, version, "compile", "jar", null,
+                    DefaultArtifactHandler("jar"))
+        }
 
         log.debug("Resolving $groupId:$artifactId:$version ...")
 
         val result: ArtifactResolutionResult = repositorySystem!!.resolve(request)
-        resolutionErrorHandler!!.throwErrors(request, result);
-        return result.artifacts.stream().map { it.file }.collect(Collectors.toList())
+        resolutionErrorHandler!!.throwErrors(request, result)
+        return result.artifacts.map { it.file }
     }
 
     private val dokkaVersion: String by lazy {

--- a/runners/maven-plugin/src/main/kotlin/DokkaMojo.kt
+++ b/runners/maven-plugin/src/main/kotlin/DokkaMojo.kt
@@ -6,6 +6,7 @@ import org.apache.maven.artifact.DefaultArtifact
 import org.apache.maven.artifact.handler.DefaultArtifactHandler
 import org.apache.maven.artifact.resolver.ArtifactResolutionRequest
 import org.apache.maven.artifact.resolver.ArtifactResolutionResult
+import org.apache.maven.artifact.resolver.ResolutionErrorHandler
 import org.apache.maven.execution.MavenSession
 import org.apache.maven.model.Dependency
 import org.apache.maven.plugin.AbstractMojo
@@ -59,6 +60,9 @@ abstract class AbstractDokkaMojo(private val defaultDokkaPlugins: List<Dependenc
 
     @Component
     private var repositorySystem: RepositorySystem? = null;
+
+    @Component
+    private var resolutionErrorHandler: ResolutionErrorHandler? = null;
 
     class PackageOptions : DokkaConfiguration.PackageOptions {
         @Parameter
@@ -237,8 +241,8 @@ abstract class AbstractDokkaMojo(private val defaultDokkaPlugins: List<Dependenc
             offlineMode = offlineMode,
             cacheRoot = cacheRoot?.let(::File),
             sourceSets = listOf(sourceSet),
-            pluginsClasspath = getArtifactByAether("org.jetbrains.dokka", "dokka-base", dokkaVersion) +
-                    dokkaPlugins.map { getArtifactByAether(it.groupId, it.artifactId, it.version ?: dokkaVersion) }
+            pluginsClasspath = getArtifactByMaven("org.jetbrains.dokka", "dokka-base", dokkaVersion) +
+                    dokkaPlugins.map { getArtifactByMaven(it.groupId, it.artifactId, it.version ?: dokkaVersion) }
                         .flatten(),
             pluginsConfiguration = pluginsConfiguration.toMutableList(),
             modules = emptyList(),
@@ -250,13 +254,13 @@ abstract class AbstractDokkaMojo(private val defaultDokkaPlugins: List<Dependenc
         gen.generate()
     }
 
-    private fun getArtifactByAether(
+    private fun getArtifactByMaven(
         groupId: String,
         artifactId: String,
         version: String
     ): List<File> {
 
-        val request: ArtifactResolutionRequest = ArtifactResolutionRequest()
+        val request = ArtifactResolutionRequest()
         request.isResolveRoot = true
         request.isResolveTransitively = true
         request.localRepository = session!!.localRepository
@@ -269,7 +273,10 @@ abstract class AbstractDokkaMojo(private val defaultDokkaPlugins: List<Dependenc
         request.artifact = DefaultArtifact(groupId, artifactId, version, "compile", "jar", null,
                 DefaultArtifactHandler("jar"))
 
+        log.debug("Resolving $groupId:$artifactId:$version ...")
+
         val result: ArtifactResolutionResult = repositorySystem!!.resolve(request)
+        resolutionErrorHandler!!.throwErrors(request, result);
         return result.artifacts.stream().map { it.file }.collect(Collectors.toList())
     }
 

